### PR TITLE
Add unit tests

### PR DESCRIPTION
**Pull request by olafurpg**
This PR brings in the ping-pong tests from the unmerged PR #4 and also BaseProtocolMessageSuite from the metals repo. While working on the unit tests, I found some low-hanging fruit improvements to the API to simplify management of input/output pairs.

cc/ @andyscott

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -12,14 +12,11 @@ jobs:
...
-        - sbt ++2.11.12 jsonrpc/test
-    - env: TEST="sbt test"
-      script:
-        - sbt test
+      script: sbt ++2.11.12 test
+    - env: TEST="2.12"
+      script: sbt test
     - stage: release
-      script: sbt ci-release
+      script: travis_wait 60 sbt ci-release
```
Comment (gabro) [inline .travis.yml:19]: Why is this wait needed?

**Review by gabro — APPROVED** [review]
LGTM, I like the new API too. I have a minor question about Travis.

File: .travis.yml (hunk 1 above)
Comment (olafurpg) [reply .travis.yml:19]: sonatypeReleaseAll can be so slow that Travis kills the release job due to "no activity". Happened last week in the lsp4s repo.

Comment (gabro) [reply .travis.yml:19]: Ok! I misunderstood what the function does: it doesn't wait, it sets a different timeout. 👍

**Review by jvican — APPROVED** [review]
LGTM 👍

**Pull request merged**